test/maybe.test.js

```diff
@@ -0,0 +1,180 @@
+const maybe = require("monio/maybe");
+const just = require("monio/just");
+const qunit = require("qunit");
+const { identity, inc, twice, justProp, maybeProp } = require("./utils");
+
+qunit.module("maybe");
+
+qunit.test("#unit", (assert) => {
+	assert.equal(
+		maybe(1)._inspect(),
+		"Maybe:Just(1)",
+		"should create a Maybe monad via monad"
+	);
+
+	assert.equal(
+		maybe.of(1)._inspect(),
+		"Maybe:Just(1)",
+		"should create a Maybe monad via #of"
+	);
+
+	assert.equal(
+		maybe.pure(1)._inspect(),
+		"Maybe:Just(1)",
+		"should create a Maybe monad via #pure"
+	);
+
+	assert.equal(
+		maybe.unit(1)._inspect(),
+		"Maybe:Just(1)",
+		"should create a Maybe monad via #unit"
+	);
+
+	assert.equal(
+		maybe.Just(1)._inspect(),
+		maybe(1)._inspect(),
+		"should create a Maybe Just monad via #Just"
+	);
+
+	assert.equal(
+		maybe.of(just.of(1))._inspect(),
+		maybe.of(1)._inspect(),
+		"should create a Maybe Just monad from a Just value"
+	);
+});
+
+qunit.test("#Nothing", (assert) => {
+	assert.equal(
+		maybe.Nothing()._inspect(),
+		"Maybe:Nothing()",
+		"should create a Maybe Nothing"
+	);
+});
+
+qunit.test("#map", (assert) => {
+	assert.equal(
+		maybe.of(1).map(identity)._inspect(),
+		maybe.of(1)._inspect(),
+		"should follow the identity law for a Maybe Just monad"
+	);
+
+	assert.equal(
+		maybe.of(1).map(inc).map(twice)._inspect(),
+		maybe.of(1).map(val => twice(inc(val)))._inspect(),
+		"should follow the composition law for a Maybe Just monad"
+	);
+
+	assert.equal(
+		maybe.Nothing().map(inc)._inspect(),
+		maybe.Nothing()._inspect(),
+		"should perform no operation on a Maybe Nothing monad"
+	);
+});
+
+qunit.test("#chain", (assert) => {
+	assert.equal(
+		maybe({ name: "john" }).chain(maybeProp("name"))._inspect(),
+		maybe.Just("john")._inspect(),
+		"should return a Maybe Just monad with 'john' as value"
+	);
+
+	assert.equal(
+		maybe.Nothing().chain(maybeProp("name"))._inspect(),
```

Comment:
Do you think we should add extra tests to *verify* that the short-circuiting is happening (especially for like the Either's coming up)? This test assumes that since the resulting monad is still a *Nothing*, the function passed to `chain(..)` had no effect. But should we have a test where we put a spy function there to prove it doesn't even get called? Is that level of verification important?

Reply:
Never thought about it. Lemme write some tests to get a better grasp on it, and I'll let you know what I think :)

Reply:
basically, you (*ironically*) have to create impure test conditions to verify that the FP stuff isn't doing impure stuff. :)

Reply:
I don't know if `sinon` is the best option for this, but it's the one that I have experience with